gpflux/layers/gi_gp_layer.py

```diff
@@ -0,0 +1,372 @@
+
+        if inducing_targets is None:
+            inducing_targets = np.zeros((self.num_latent_gps, num_inducing, 1))
+        elif tf.rank(inducing_targets) == 2:
+            inducing_targets = tf.expand_dims(tf.linalg.adjoint(inducing_targets), -1)
+        if inducing_targets.shape != (self.num_latent_gps, num_inducing, 1):
+            raise ValueError("Incorrect shape was provided for the inducing targets.")
+
+        self.v = Parameter(
+            inducing_targets,
+            dtype=default_float(),
+            name=f"{self.name}_v" if self.name else "v",
+        )
+
+        self.L_loc = Parameter(
+            np.stack([np.eye(num_inducing) for _ in range(self.num_latent_gps)]),
+            transform=triangular(),
+            dtype=default_float(),
+            name=f"{self.name}_L_loc" if self.name else "L_loc",
+        )  # [num_latent_gps, num_inducing, num_inducing]
+
+        self.L_scale = Parameter(
+            tf.sqrt(self.kernel.variance)*np.sqrt(prec_init)*np.ones((self.num_latent_gps, 1, 1)),
+            transform=positive(),
+            dtype=default_float(),
+            name=f"{self.name}_L_scale" if self.name else "L_scale"
+        )
+
+    @property
+    def L(self) -> tf.Tensor:
+        """
+        :return: the Cholesky of the precision hyperparameter. We parameterize L using L_loc and
+            L_scale to achieve greater stability during optimization.
+        """
+        norm = tf.reshape(tf.reduce_mean(tf.linalg.diag_part(self.L_loc), axis=-1), [-1, 1, 1])
+        return self.L_loc * self.L_scale/norm
+
+    def mvnormal_log_prob(self, sigma_L: TensorType, X: TensorType) -> tf.Tensor:
+        """
+        Calculates the log probability of a zero-mean multivariate Gaussian with covariance sigma
+        and evaluation points X, with batching of both the covariance and X.
+
+        TODO: look into whether this can be replaced with a tfp.distributions.Distribution
+        :param sigma_L: Cholesky of covariance sigma, shape [..., 1, D, D]
+        :param X: evaluation point for log_prob, shape [..., M, D, 1]
+        :return: the log probability, shape [..., M]
+        """
+        in_features = tf.cast(tf.shape(X)[-2], dtype=default_float())
+        out_features = tf.cast(tf.shape(X)[-1], dtype=default_float())
+        trace_quad = tf.reduce_sum(tf.linalg.triangular_solve(sigma_L, X)**2, [-1, -2])
+        logdet_term = 2.0*tf.reduce_sum(tf.math.log(tf.linalg.diag_part(sigma_L)), -1)
+        return -0.5*trace_quad - 0.5*out_features*(logdet_term + in_features*math.log(2*math.pi))
+
+    def call(
+        self,
+        inputs: TensorType,
+        *args: List[Any],
+        **kwargs: Dict[str, Any]
+    ) -> tf.Tensor:
+        """
+        Sample-based propagation of both inducing points and function values. See Ober & Aitchison
+        (2021) for details.
+        """
+        mean_function = self.mean_function(inputs)
+
+        Kuu = self.kernel(inputs[..., :self.num_inducing, :])
```

Comment:
Can you give some information in the form of comments how `inputs` is structured. What exactly are the inputs and outputs of `call`?